pydicom/charset.py

```diff
@@ -151,42 +162,166 @@ def decode_fragment(byte_str, encodings, delimiters):
+            return byte_str.decode(encoding)
+
+    # unknown escape code - use first encoding
+    warnings.warn(u"Found unknown escape sequence in encoded string value - "
+                  u"using encoding {}".format(encodings[0]))
+    return byte_str.decode(encodings[0], errors='replace')
+
+
+def encode_string(value, encodings):
+    """Convert a unicode string into a byte string using the given
+    list of encodings.
+
+    Parameters
+    ----------
+    value : text type
+        The unicode string as presented to the user.
+    encodings : list
+        The encodings needed to encode the string as a list of Python
+        encodings, converted from the encodings in Specific Character Set.
+
+    Returns
+    -------
+    byte string
+        The encoded string. If the value could not be encoded with any of
+        the given encodings, a warning is issued, and the value is
+        encoded using the first encoding with replacement characters,
+        resulting in data loss.
+    """
+    for i, encoding in enumerate(encodings):
+        try:
+            encoded = value.encode(encoding)
+            if i > 0 and encoding not in handled_encodings:
+                return ENCODINGS_TO_CODES.get(encoding, b'') + encoded
+            return encoded
+        except UnicodeError:
+            continue
+    else:
+        # if we have more than one encoding, we retry encoding by splitting
+        # `value` into chunks that can be encoded with one of the encodings
+        if len(encodings) > 1:
+            try:
+                return _encode_string_parts(value, encodings)
+            except ValueError:
+                pass
+        # all attempts failed - warn and encode with replacement characters
+        warnings.warn("Failed to encode value with encodings: {} - using "
+                      "replacement characters in encoded string"
+                      .format(', '.join(encodings)))
+        return value.encode(encodings[0], errors='replace')
+
+
+def _encode_string_parts(value, encodings):
+    """Convert a unicode string into a byte string using the given
+    list of encodings.
+    This is invoked if `encode_string` failed to encode `value` with a single
+    encoding. We try instead to use different encodings for different parts
+    of the string, using the encoding that can encode the longest part of
+    the rest of the string as we go along.
+
+    Parameters
+    ----------
+    value : text type
+        The unicode string as presented to the user.
+    encodings : list
+        The encodings needed to encode the string as a list of Python
+        encodings, converted from the encodings in Specific Character Set.
+
+    Returns
+    -------
+    byte string
+        The encoded string, including the escape sequences needed to switch
+        between different encodings.
+
+    Raises
+    ------
+    ValueError
+        If `value` could not be encoded with the given encodings.
+
+    """
+    encoded = b''
```

Comment:
This might be be better as a `bytearray` if the length of `encoded` gets large.

Reply:
Good idea! I will probably go with the byte array.